INT-3584: Introduce `FileSplitter`

artembilan opened the pull request:
JIRA: https://jira.spring.io/browse/INT-3584

artembilan committed:
INT-3584: Introduce `FileSplitter`

JIRA: https://jira.spring.io/browse/INT-3584

artembilan commented:
Let me know if it would make sence to add `What's New` note already.

garyrussell commented on spring-integration-file/src/main/java/org/springframework/integration/file/splitter/FileSplitter.java:
```diff
@@ -0,0 +1,147 @@
+
+	public void setIterator(boolean iterator) {
+		this.iterator = iterator;
+	}
```
Let's make it `final` and a constructor arg.

I also think it should be `true` by default.

To be honest, I see little use for the `false` case since we'll load the entire file into memory and then `produceOutput` iterates over the collection. While there might be a corner case where that is desired, I suspect the iterator version would be most useful.

artembilan replied:
That's true. But meanwhile I'm looking into the `XPathMessageSplitter`, which should produce `iterator` as well. And if we make it as an `iterator` by default it might produce some side effect. However we always can point that out in the Migration Guide...

garyrussell replied:
Just because true is the default in the file splitter doesn't mean we can't have false default for the XPathSplitter.

artembilan replied:
Yes, we can, but it won't be consistent throughout the project and from end-application perspective.
That's why I say that do that `iterator="true"` for the `XPathSplitter` in the 4.2 would be good choice.
And yes, I'm coming back to the `iterator="true"` for the `FileSplitter`

artembilan committed:
INT-3584: Move `iterator` option to ctor

artembilan commented:
Pushed.
Created https://jira.spring.io/browse/INT-3599 for the `XPathMessageSplitter` planed for 4.2

garyrussell committed:
Polishing

- Javadocs
- Fix String payload (missing 'else')
- Support charsets other than the default
- throw MessagingException instead of RuntimeException
- Add more tests to cover all cases

garyrussell commented:
LGTM; polishing here: https://github.com/garyrussell/spring-integration/commit/d7f9ecfb4f0d48f7eb2bda392bb9962d0ed2f30f

artembilan committed:
INT-3584: Throw `MessageHandlingException`

garyrussell commented:
Merged as 703e69ff5c6988b5a0057439e6582e2edaf5c507
